### PR TITLE
Metric: MAD

### PR DESCRIPTION
**Pull request by Shamus98**
**IMPORTANT: Please do not create a Pull Request without creating an issue first.**

*Any change needs to be discussed before proceeding. Failure to do so may result in the rejection of the pull request.
This text need to be deleted before submitting the PR.*

Closes #

## Proposed Changes

  -
  -
  -

### CODE REVIEW

File: tests/test_mad.py [hunk 1]
```diff
@@ -0,0 +1,577 @@
...
+    https://colab.research.google.com/drive/10W6-JLI26FsVsrpolWGG9Gb7YjDoS3hy
+"""
+
+import numpy as np
+import torch
+from torchvision import transforms
+import torch.nn as nn
+import torch.nn.functional as F
+import math
+import functools
```
Comment (zakajd) [inline tests/test_mad.py:16]: Remove unnecessary imports (numpy, torchvision, math, etc.)

File: tests/test_mad.py [hunk 2]
```diff
@@ -0,0 +1,577 @@
...
+def abs(x):
+    return torch.sqrt(x[:,:,:,:,0]**2+x[:,:,:,:,1]**2+1e-12)
+
+def real(x):
+    return x[:,:,:,:,0]
+
+def imag(x):
+    return x[:,:,:,:,1]
+
+def roll_n(X, axis, n):
```
Comment (zakajd) [inline tests/test_mad.py:29]: Use `torch.roll`

File: tests/test_mad.py [hunk 3]
```diff
@@ -0,0 +1,577 @@
...
+		# [0, 100] => [-1, 1],  ~[-110, 110] => [-1, 1]
+		return [L_chan / 50.0 - 1.0, a_chan / 110.0, b_chan / 110.0]
+
+def deprocess_lab(L_chan, a_chan, b_chan):
+		#TODO This is axis=3 instead of axis=2 when deprocessing batch of images 
+			   # ( we process individual images but deprocess batches)
+		#return tf.stack([(L_chan + 1) / 2 * 100, a_chan * 110, b_chan * 110], axis=3)
+		return torch.stack([(L_chan + 1) / 2.0 * 100.0, a_chan * 110.0, b_chan * 110.0], dim=2)
+
+def rgb_to_lab(srgb):
```
Comment (zakajd) [inline tests/test_mad.py:66]: We have colour conversion functions in `piq.functonal`.  Check it out and remove this functions along with `preprocess_lab` and `deprocess_lab`

File: tests/test_mad.py [hunk 4]
```diff
@@ -0,0 +1,577 @@
...
+        #  l       a       b
+        [  0.0,  500.0,    0.0], # fx
+        [116.0, -500.0,  200.0], # fy
+        [  0.0,    0.0, -200.0], # fz
+    ]).type(torch.FloatTensor).to(device)
+    lab_pixels = torch.mm(fxfyfz_pixels, fxfyfz_to_lab) + torch.tensor([-16.0, 0.0, 0.0]).type(torch.FloatTensor).to(device)
+    #return tf.reshape(lab_pixels, tf.shape(srgb))
+    return torch.reshape(lab_pixels, srgb.shape)
+
+def lab_to_rgb(lab):
```
Comment (zakajd) [inline tests/test_mad.py:101]: This function is never used

File: tests/test_mad.py [hunk 5]
```diff
@@ -0,0 +1,577 @@
...
+		srgb_pixels = (rgb_pixels * 12.92 * linear_mask) + (((rgb_pixels+0.000001) ** (1/2.4) * 1.055) - 0.055) * exponential_mask
+	
+		return torch.reshape(srgb_pixels, lab.shape)
+
+def spatial_normalize(x):
+    min_v = torch.min(x.view(x.shape[0],1,-1),dim=2)[0]
+    range_v = torch.max(x.view(x.shape[0],1,-1),dim=2)[0] - min_v
+    return (x - min_v.unsqueeze(2).unsqueeze(3)) / (range_v.unsqueeze(2).unsqueeze(3)+1e-12)
+
+def fspecial_gauss(size, sigma, channels):
```
Comment (zakajd) [inline tests/test_mad.py:148]: See https://github.com/photosynthesis-team/piq/blob/master/piq/functional/filters.py#L27

File: tests/test_mad.py [hunk 6]
```diff
@@ -0,0 +1,577 @@
...
+    # Function to mimic the 'fspecial' gaussian MATLAB function
+    x, y = np.mgrid[-size//2 + 1:size//2 + 1, -size//2 + 1:size//2 + 1]
+    g = np.exp(-((x**2 + y**2)/(2.0*sigma**2)))
+    g = torch.from_numpy(g/g.sum()).float().unsqueeze(0).unsqueeze(0)
+    return g.repeat(channels,1,1,1)
+
+def downsample(img1, img2, maxSize = 256):
```
Comment (zakajd) [inline tests/test_mad.py:155]: `torch.nn.functional.avgpool2d()`

File: tests/test_mad.py [hunk 7]
```diff
@@ -0,0 +1,577 @@
...
+    _,channels,H,W = img1.shape
+    f = int(max(1,np.round(min(H,W)/maxSize)))
+    if f>1:
+        aveKernel = (torch.ones(channels,1,f,f)/f**2).to(img1.device)
+        img1 = F.conv2d(img1, aveKernel, stride=f, padding = 0, groups = channels)
+        img2 = F.conv2d(img2, aveKernel, stride=f, padding = 0, groups = channels)
+    return img1, img2
+
+def extract_patches_2d(img, patch_shape=[64, 64], step=[27,27], batch_first=True, keep_last_patch=False):
```
Comment (zakajd) [inline tests/test_mad.py:164]: See https://github.com/photosynthesis-team/piq/blob/06c83b7ed317cd2ea2a78b27ecc858681f99e69b/piq/functional/base.py#L89

File: tests/test_mad.py [hunk 8]
```diff
@@ -0,0 +1,577 @@
...
+        patches = patches.permute(1,0,2,3,4)
+    return patches.reshape(-1,3,patch_H,patch_W)
+
+def prepare_image(image, resize = False, repeatNum = 1):
+    if resize and min(image.size)>256:
+        image = transforms.functional.resize(image,256)
+    image = transforms.ToTensor()(image)
+    return image.unsqueeze(0).repeat(repeatNum,1,1,1)
+
+def print_network(net):
```
Comment (zakajd) [inline tests/test_mad.py:197]: This function is never used

File: tests/test_mad.py [hunk 9]
```diff
@@ -0,0 +1,577 @@
...
+    patches_fold_HW = patches_fold_H.unfold(3, patch_W, step_int[1])   
+    if((img.size(3) - patch_W) % step_int[1] != 0) and keep_last_patch:
+        patches_fold_HW = torch.cat((patches_fold_HW,patches_fold_H[:,:,:,-patch_W:,:].permute(0,1,2,4,3).unsqueeze(3)),dim=3)
+    patches = patches_fold_HW.permute(2,3,0,1,4,5)
+    patches = patches.reshape(-1,img.size(0),img.size(1),patch_H,patch_W)
+    if(batch_first):
+        patches = patches.permute(1,0,2,3,4)
+    return patches.reshape(-1,3,patch_H,patch_W)
+
+def prepare_image(image, resize = False, repeatNum = 1):
```
Comment (zakajd) [inline tests/test_mad.py:191]: Move resize to main function

File: tests/test_mad.py [hunk 10]
```diff
@@ -0,0 +1,577 @@
...
+    return image.unsqueeze(0).repeat(repeatNum,1,1,1)
+
+def print_network(net):
+    num_params = 0
+    for param in net.parameters():
+        num_params += param.numel()
+    print(net)
+    print('Total number of parameters: %d' % num_params)
+
+from typing import Optional, Union, Tuple, List, Iterable
```
Comment (zakajd) [inline tests/test_mad.py:204]: All imports should be on top.
Install `flake8` and run it on your code:
```
pip install flake8
flake8 .
```

File: tests/test_mad.py [hunk 11]
```diff
@@ -0,0 +1,577 @@
...
+import torch
+
+
+def _adjust_dimensions(input_tensors: Union[torch.Tensor, Iterable[torch.Tensor]]):
```
Comment (zakajd) [inline tests/test_mad.py:208]: Duplicates `piq.common._adjust_dimensions`

File: tests/test_mad.py [hunk 12]
```diff
@@ -0,0 +1,577 @@
...
+            raise ValueError(f'Expected 2, 3, 4 or 5 dimensions (got {tensor.dim()})')
+        resized_tensors.append(tmp)
+
+    if len(resized_tensors) == 1:
+        return resized_tensors[0]
+
+    return tuple(resized_tensors)
+
+
+def _validate_input(
```
Comment (zakajd) [inline tests/test_mad.py:231]: Duplicates `piq.common._valiadate_input`

File: tests/test_mad.py [hunk 13]
```diff
@@ -0,0 +1,577 @@
...
+    if scale_weights is not None:
+        assert isinstance(scale_weights, (list, tuple, torch.Tensor)), \
+            f'Scale weights must be of type list, tuple or torch.Tensor, got {type(scale_weights)}.'
+        if isinstance(scale_weights, (list, tuple)):
+            scale_weights = torch.tensor(scale_weights)
+        assert (scale_weights.dim() == 1), \
+            f'Scale weights must be one dimensional, got {scale_weights.dim()}.'
+
+
+def _validate_features(x: torch.Tensor, y: torch.Tensor) -> None:
```
Comment (zakajd) [inline tests/test_mad.py:274]: Not used

File: tests/test_mad.py [hunk 14]
```diff
@@ -0,0 +1,577 @@
...
+import numpy as np
+import os
+import sys
+import torch
+from torchvision import models,transforms
+import torch.nn as nn
+import torch.nn.functional as F
+import inspect
+from numpy.fft import fft2, ifft2, fftshift, ifftshift
+import math
```
Comment (zakajd) [inline tests/test_mad.py:300]: Move on top and delete unnecessary  imports

File: tests/test_mad.py [hunk 15]
```diff
@@ -0,0 +1,577 @@
...
+#from utils import abs, real, imag, downsample, batch_fftshift2d, batch_ifftshift2d
+
+MAX = nn.MaxPool2d((2,2), stride=1, padding=1)
+
+def extract_patches_2d(img, patch_shape=[64, 64], step=[27,27], batch_first=False, keep_last_patch=False):
```
Comment (zakajd) [inline tests/test_mad.py:305]: Duplicate

File: tests/test_mad.py [hunk 16]
```diff
@@ -0,0 +1,577 @@
...
+            mp = mp + s[gb_i] * ( torch.abs( stdref - stddst ) + 2*torch.abs( skwref - skwdst ) +  torch.abs( krtref - krtdst ) ) 
+
+    # mp2 = mp[:,:, BSIZE+1:-BSIZE-1, BSIZE+1:-BSIZE-1]
+    B, C, rows, cols = mp.shape
+    return torch.norm( mp.reshape(B,C,-1) , dim=2 ) / np.sqrt(rows * cols) 
+
+
+def mad(x:torch.Tensor, y: torch.Tensor, reduction: str = 'mean',
+            data_range: Union[int, float] = 1., scales: int = 3, subsample: bool = True,
+            c: float = 30.0, alpha: float = 4.2) -> torch.Tensor:
```
Comment (zakajd) [inline tests/test_mad.py:526]: `c` and `alpha`variables are not used. Add `thresh1`, `thresh2` and others instead to make function more flexible.

File: tests/test_mad.py [hunk 17]
```diff
@@ -0,0 +1,577 @@
...
+        self.data_range = data_range
+        self.mad = functools.partial(
+            mad, scales=scales, subsample=subsample, c=c, alpha=alpha,
+            data_range=data_range, reduction=reduction)
+
+    def forward(self, prediction, target):
+        return 1. - self.mad(x=prediction, y=target)
+
+
+if __name__ == '__main__':
```
Comment (zakajd) [inline tests/test_mad.py:560]: Delete this

File: tests/test_mad.py [hunk 18]
```diff
@@ -0,0 +1,577 @@
+# -*- coding: utf-8 -*-
```
Comment (zakajd) [inline tests/test_mad.py:1]: This file doesn't contaitn tests. It should be renamed `mad.py` and placed into `piq.piq` folder.

**Review by zakajd — CHANGES_REQUESTED** [review]
Thanks for opening PR.
Overall it's in a VERY raw stage. 
Please address this comments and then let's continue the discussion

**Comment by zakajd** [comment]
Temporary closed after discussion with @Shamus98
